File: src/main/java/magma/app/FieldTranspiler.java
package magma.app;

import java.util.Arrays;

class FieldTranspiler {
    static String transpileFields(String source) {
        var lines = source.split("\\R");
        var out = new StringBuilder();
        for (var line : lines) {
            var trimmed = line.trim();
            if (!trimmed.endsWith(";") || trimmed.contains("(") || trimmed.startsWith("import") ||
                trimmed.startsWith("return") || trimmed.startsWith("let ")) {
                out.append(line).append(System.lineSeparator());
                continue;
            }

            var indent = line.substring(0, line.indexOf(trimmed));
            var withoutSemi = trimmed.substring(0, trimmed.length() - 1).trim();
            var eq = withoutSemi.indexOf('=');
            if (eq != -1) {
                withoutSemi = withoutSemi.substring(0, eq).trim();
            }
            var tokens = withoutSemi.split("\\s+");
            if (tokens.length < 2) {
                out.append(line).append(System.lineSeparator());
                continue;
            }

            var name = tokens[tokens.length - 1];
            var type = tokens[tokens.length - 2];
            var modArray = Arrays.copyOf(tokens, tokens.length - 2);
            var modifiers = replaceFinalWithReadonly(modArray);
            var tsType = TypeMapper.toTsType(type);
            out.append(indent);
            if (!modifiers.isBlank()) {
                out.append(modifiers).append(" ");
            }
            out.append(name).append(": ").append(tsType).append(";")
               .append(System.lineSeparator());
        }
        return out.toString().trim();
    }

    private static String replaceFinalWithReadonly(String[] mods) {
        for (var i = 0; i < mods.length; i++) {
            if (mods[i].equals("final")) {
                mods[i] = "readonly";
            }
        }
        return String.join(" ", mods).trim();
    }
}

File: src/main/node/magma/app/FieldTranspiler.ts
import Arrays from "../../java/util/Arrays";
export default class FieldTranspiler {
    transpileFields(source: string): string {
        let lines : unknown = source.split("\\R");
        let out : StringBuilder = new StringBuilder();
        // TODO
        let trimmed : unknown = line.trim();
        // TODO
        ") || trimmed.contains("(") || trimmed.startsWith("import");
        /* */: TODO;
        out.append(line).append(System.lineSeparator());
        // TODO
        // TODO
        let indent : unknown = line.substring(0, line.indexOf(trimmed));
        let withoutSemi : unknown = trimmed.substring(0, trimmed.length()).trim();
        let eq : unknown = withoutSemi.indexOf(/* TODO */);
        if (/* TODO */) {
            // TODO
        }
        let tokens : unknown = withoutSemi.split("\\s+");
        if (tokens.length < 2) {
            out.append(line).append(System.lineSeparator());
            // TODO
        }
        let name : unknown = tokens[tokens.length - 1];
        let type : unknown = tokens[tokens.length - 2];
        let modArray : unknown = Arrays.copyOf(tokens, tokens.length - 2);
        let modifiers : unknown = replaceFinalWithReadonly(modArray);
        let tsType : unknown = TypeMapper.toTsType(type);
        out.append(indent);
        if (!modifiers.isBlank()) {
            out.append(modifiers).append(" ");
        }
        out.append(name).append(": ").append(tsType);
        // TODO
        .append(System.lineSeparator());
        // TODO
        return out.toString().trim();
    }

    replaceFinalWithReadonly(mods: string[]): string {
        let i : (var = 0;
        i mods.length: <;
        // TODO
        if (mods[i].equals("final")) {
            // TODO
        }
        // TODO
        return String.join(" ", mods).trim();
    }
}
